### PR TITLE
Add hybrid quantization for StableDiffusion pipelines

### PR DESCRIPTION
**Pull request by l-bat**
# What does this PR do?

* Enabled quantization in hybrid mode for `OVStableDiffusionPipeline`, `OVStableDiffusionXLPipeline` and `OVLatentConsistencyModelPipeline`, when the part of the model is fully quantized and the weights of another part are just compressed
* Added set of pre-defined datasets: `['conceptual_captions','laion/220k-GPT4Vision-captions-from-LIVIS','laion/filtered-wit']`
* Usage example
```python
from optimum.intel.openvino import OVStableDiffusionPipeline
from optimum.intel import OVWeightQuantizationConfig

model_id = "stabilityai/stable-diffusion-2-1" 
quantization_config = OVWeightQuantizationConfig(bits=8, dataset="conceptual_captions", subset_size=200)
model = OVStableDiffusionPipeline.from_pretrained(model_id, export=True, quantization_config=quantization_config)
```

## Before submitting
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/intel/openvino/quantization.py [hunk 1]
```diff
@@ -584,4 +588,98 @@ def _weight_only_quantization(
...
+
+
+def _hybrid_quantization(model: openvino.runtime.Model, quantization_config: Union[OVWeightQuantizationConfig, Dict]):
+    dataset = quantization_config.dataset
+    wc_ignored_scope = deepcopy(quantization_config.ignored_scope)
+
+    if isinstance(wc_ignored_scope, dict):
+        wc_ignored_scope["types"] = wc_ignored_scope.get("types", []) + ["Convolution"]
+    else:
+        assert wc_ignored_scope is None
```
Comment (echarlaix) [inline optimum/intel/openvino/quantization.py:662]: would prefer to raise an error here (with explicit message)

File: optimum/intel/openvino/quantization.py [hunk 2]
```diff
@@ -584,4 +588,98 @@ def _weight_only_quantization(
...
+    compressed_model = _weight_only_quantization(model, quantization_config)
+
+    quantized_model = nncf.quantize(
+        compressed_model,
+        dataset,
+        model_type=nncf.ModelType.TRANSFORMER,
+        ignored_scope=nncf.IgnoredScope(**ptq_ignored_scope),
+        advanced_parameters=nncf.AdvancedQuantizationParameters(AdvancedSmoothQuantParameters(matmul=-1)),
+        subset_size=quantization_config.subset_size,
+    )
```
Comment (echarlaix) [inline optimum/intel/openvino/quantization.py:684]: why do we need to do this ? could you explain a bit by adding comments + explaining what is meant by hybrid quantization ?

**Review by echarlaix — COMMENTED** [review]
Thanks a lot for your work @l-bat

File: optimum/intel/openvino/modeling_diffusion.py [hunk 3]
```diff
@@ -291,6 +301,36 @@ def _from_pretrained(
...
+            supported_pipelines = (
+                OVStableDiffusionPipeline,
+                OVStableDiffusionXLPipeline,
+                OVLatentConsistencyModelPipeline,
+            )
+            if not isinstance(sd_model, supported_pipelines):
+                raise NotImplementedError(f"Quantization in hybrid mode is not supported for {cls.__name__}")
+
+            num_inference_steps = 4 if isinstance(cls, OVLatentConsistencyModelPipeline) else 50
+            quantization_config.dataset = dataset
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:316]: Here you modify the value in the `quantization_config` object that the user passed to the API. This is not good, IMO. I would recommend passing `dataset` explicitly to `_hybrid_quantization()`.

File: optimum/intel/openvino/modeling_diffusion.py [hunk 4]
```diff
@@ -274,9 +276,17 @@ def _from_pretrained(
...
 
         quantization_config = cls._prepare_weight_quantization_config(quantization_config, load_in_8bit)
-        unet = cls.load_model(
-            new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name, quantization_config
-        )
+
+        dataset = None
+        if quantization_config:
+            dataset = quantization_config.dataset
+            quantization_config.dataset = None  # apply weight compression without dataset
```
[the remark this answers is not in the document]
Comment (l-bat) [reply optimum/intel/openvino/modeling_diffusion.py]: removed

File: optimum/intel/openvino/modeling_diffusion.py [hunk 5]
```diff
@@ -291,6 +301,36 @@ def _from_pretrained(
         if model_save_dir is None:
             model_save_dir = new_model_save_dir
 
+        if quantization_config and dataset is not None:
```
[the remark this answers is not in the document]
Comment (l-bat) [reply optimum/intel/openvino/modeling_diffusion.py]: removed

File: optimum/intel/openvino/configuration.py [hunk 6]
```diff
@@ -219,6 +223,7 @@ def __init__(
         self.all_layers = all_layers
         self.sensitivity_metric = sensitivity_metric
         self.ignored_scope = ignored_scope
+        self.subset_size = subset_size
```
[the remark this answers is not in the document]
Comment (l-bat) [reply optimum/intel/openvino/configuration.py]: Added

File: optimum/intel/openvino/quantization.py (hunk 1 above)
Comment (l-bat) [reply optimum/intel/openvino/quantization.py]: I reformatted this part

File: optimum/intel/openvino/quantization.py (hunk 2 above)
Comment (l-bat) [reply optimum/intel/openvino/quantization.py]: Added docstring to `_hybrid_quantization`.

On the one hand, post-training quantization of the UNet model requires leads to accuracy drop. On the other hand, the weight compression doesn't improve performance when applying to Stable Diffusion models, because the size of activations is comparable to weights. That is why the proposal is to apply quantization in hybrid mode which means that we quantize: (1) weights of MatMul and Embedding layers and (2) activations of other layers.

File: optimum/intel/openvino/modeling_diffusion.py (hunk 3 above)
Comment (l-bat) [reply optimum/intel/openvino/modeling_diffusion.py]: created deepcopy of `quantization_config`  to pass it to `_hybrid_quantization`.

Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_diffusion.py]: I don't think it is safe enough to create a deep copy. Imagine someone passes a huge dataset inside the config. We will copy it as far as I understand.

**Comment by helena-intel** [comment]
This looks very interesting, thanks @l-bat ! Can you add this to the documentation? https://github.com/huggingface/optimum-intel/blob/main/docs/source/optimization_ov.mdx It would also be good to add a note with a link to this example https://github.com/huggingface/optimum-intel/tree/main/examples/openvino/stable-diffusion (for people who find that when looking for quantizing stable diffusion, to let them know about this option)

File: docs/source/optimization_ov.mdx [hunk 7]
```diff
@@ -69,6 +69,23 @@ from optimum.intel import OVModelForCausalLM
 model = OVModelForCausalLM.from_pretrained(model_id, load_in_8bit=True)
 ```
 
+##  Hybrid quantization
+
+Traditional optimization methods like post-training 8-bit quantization do not work for Stable Diffusion models because accuracy drops significantly. On the other hand, weight compression does not improve performance when applied to Stable Diffusion models, as the size of activations is comparable to weights.
```
Comment (AlexKoff88) [inline docs/source/optimization_ov.mdx:74]:
```suggestion
Traditional optimization methods like post-training 8-bit quantization do not work well for Stable Diffusion models and can lead to poor generation results. On the other hand, weight compression does not improve performance significantly when applied to Stable Diffusion models, as the size of activations is comparable to weights.
```

File: docs/source/optimization_ov.mdx [hunk 8]
```diff
@@ -69,6 +69,23 @@ from optimum.intel import OVModelForCausalLM
...
+The UNet model takes up most of the overall execution time of the pipeline. Thus, optimizing just one model brings substantial benefits in terms of inference speed while keeping acceptable accuracy without fine-tuning. Quantizing the rest of the diffusion pipeline does not significantly improve inference performance but could potentially lead to substantial degradation of accuracy.
+Therefore, the proposal is to apply quantization in hybrid mode for the UNet model and weight-only quantization for other pipeline components. The hybrid mode involves the quantization of weights in MatMul and Embedding layers, and activations of other layers, facilitating accuracy preservation post-optimization while reducing the model size.
```
Comment (AlexKoff88) [inline docs/source/optimization_ov.mdx:76]:
```suggestion
Therefore, the proposal is to apply quantization in *hybrid mode* for the UNet model and weight-only quantization for the rest of the pipeline components. The hybrid mode involves the quantization of weights in MatMul and Embedding layers, and activations of other layers, facilitating accuracy preservation post-optimization while reducing the model size.
```

File: docs/source/optimization_ov.mdx [hunk 9]
```diff
@@ -69,6 +69,23 @@ from optimum.intel import OVModelForCausalLM
...
+For optimizing the Stable Diffusion pipeline, utilize the `quantization_config` to define optimization parameters. To enable hybrid quantization, specify the quantization dataset in the `quantization_config`; otherwise, weight-only quantization in specified precisions will be applied to UNet.
```
Comment (AlexKoff88) [inline docs/source/optimization_ov.mdx:77]:
```suggestion
The `quantization_config` is utilized to define optimization parameters for optimizing the Stable Diffusion pipeline. To enable hybrid quantization, specify the quantization dataset in the `quantization_config`. Otherwise, weight-only quantization to a specified data type (8 tr 4 bits) is applied to UNet model.
```

File: optimum/intel/openvino/configuration.py [hunk 10]
```diff
@@ -177,23 +177,24 @@ class OVWeightQuantizationConfig(QuantizationConfigMixin):
...
                     using the [`~PreTrainedTokenizer.save_pretrained`] method, e.g., `./my_model_directory/`.
-        dataset (`Union[List[str]]`, *optional*):
-            The dataset used for data-aware compression. You can provide your own dataset in a list of string or just use the
-            the one from the list ['wikitext2','c4','c4-new','ptb','ptb-new']
-        group_size (`int`, *optional*, defaults to 128):
-            The group size to use for quantization. Recommended value is 128 and -1 uses per-column quantization.
-        ratio (`float`, *optional*, defaults to 1.0):
+        dataset (`str or List[str]`, *optional*):
+            The dataset used for data-aware compression or quantization with NNCF. You can provide your own dataset
+            in a list of string or just use the the one from the list ['wikitext2','c4','c4-new','ptb','ptb-new'] for LLLMs
```
Comment (AlexKoff88) [inline optimum/intel/openvino/configuration.py:182]:
```suggestion
            in a list of strings or just use the one from the list ['wikitext2','c4','c4-new','ptb','ptb-new'] for LLLMs
```

File: optimum/intel/openvino/configuration.py [hunk 11]
```diff
@@ -177,23 +177,24 @@ class OVWeightQuantizationConfig(QuantizationConfigMixin):
...
+            or ['conceptual_captions','laion/220k-GPT4Vision-captions-from-LIVIS','laion/filtered-wit'] for SD models.
```
Comment (AlexKoff88) [inline optimum/intel/openvino/configuration.py:183]:
```suggestion
            or ['conceptual_captions','laion/220k-GPT4Vision-captions-from-LIVIS','laion/filtered-wit'] for diffusion models.
```

File: optimum/intel/openvino/configuration.py [hunk 12]
```diff
@@ -231,10 +234,16 @@ def post_init(self):
...
+            stable_diffusion_datasets = [
+                "conceptual_captions",
+                "laion/220k-GPT4Vision-captions-from-LIVIS",
+                "laion/filtered-wit",
+            ]
+            if self.dataset not in llm_datasets + stable_diffusion_datasets:
                 raise ValueError(
                     f"""You have entered a string value for dataset. You can only choose between
-                    ['wikitext2','c4','c4-new','ptb','ptb-new'], but we found {self.dataset}"""
+                    {llm_datasets} for LLLMs or {stable_diffusion_datasets} for SD models, but we found {self.dataset}"""
```
Comment (AlexKoff88) [inline optimum/intel/openvino/configuration.py:246]:
```suggestion
                    {llm_datasets} for LLLMs or {stable_diffusion_datasets} for diffusion models, but we found {self.dataset}"""
```

File: optimum/intel/openvino/modeling_diffusion.py [hunk 13]
```diff
@@ -274,9 +281,17 @@ def _from_pretrained(
...
-        )
+
+        dataset = None
+        unet_path = new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name
+        if quantization_config is not None and quantization_config.dataset is not None:
+            dataset = quantization_config.dataset
+            # load the UNet model uncompressed to apply hybrid quantization further
+            unet = cls.load_model(unet_path)
+            # Apply weights compression to other `components` without dataset
+            quantization_config.dataset = None
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:292]: This is the error-prone approach when you change the values of the input argument, IMO. Please think about how we can make it more safe.

**Comment by l-bat** [comment]
> This looks very interesting, thanks @l-bat ! Can you add this to the documentation? https://github.com/huggingface/optimum-intel/blob/main/docs/source/optimization_ov.mdx It would also be good to add a note with a link to this example https://github.com/huggingface/optimum-intel/tree/main/examples/openvino/stable-diffusion (for people who find that when looking for quantizing stable diffusion, to let them know about this option)

[QAT example](https://github.com/huggingface/optimum-intel/tree/main/examples/openvino/stable-diffusion) is outdated and will be revised in the next PR. After that I will add a note with a link to this example.

File: optimum/intel/openvino/modeling_diffusion.py [hunk 14]
```diff
@@ -274,9 +281,18 @@ def _from_pretrained(
...
-            new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name, quantization_config
-        )
+
+        unet_path = new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name
+        if quantization_config is not None and quantization_config.dataset is not None:
+            # load the UNet model uncompressed to apply hybrid quantization further
+            unet = cls.load_model(unet_path)
+            # Apply weights compression to other `components` without dataset
+            q_config_params = quantization_config.__dict__
+            wc_params = {param: value for param, value in q_config_params.items() if param != "dataset"}
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:291]:
```suggestion
            weright_quantization_params = {param: value for param, value in quantization_config.__dict__.items() if param != "dataset"}
```

File: optimum/intel/openvino/modeling_diffusion.py [hunk 15]
```diff
@@ -274,9 +281,18 @@ def _from_pretrained(
...
-        unet = cls.load_model(
-            new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name, quantization_config
-        )
+
+        unet_path = new_model_save_dir / DIFFUSION_MODEL_UNET_SUBFOLDER / unet_file_name
+        if quantization_config is not None and quantization_config.dataset is not None:
+            # load the UNet model uncompressed to apply hybrid quantization further
+            unet = cls.load_model(unet_path)
+            # Apply weights compression to other `components` without dataset
+            q_config_params = quantization_config.__dict__
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_diffusion.py:290]: this string can be eliminated actually

File: optimum/intel/openvino/modeling_diffusion.py [hunk 16]
```diff
@@ -274,9 +281,18 @@ def _from_pretrained(
...
+            wc_params = {param: value for param, value in q_config_params.items() if param != "dataset"}
+            wc_quantization_config = OVWeightQuantizationConfig.from_dict(wc_params)
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:292]:
```suggestion
            weight_quantization_config = OVWeightQuantizationConfig.from_dict(wc_params)
```

File: optimum/intel/openvino/modeling_diffusion.py [hunk 17]
```diff
@@ -274,9 +281,18 @@ def _from_pretrained(
...
+        else:
+            wc_quantization_config = quantization_config
+            unet = cls.load_model(unet_path, wc_quantization_config)
```
Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:295]:
```suggestion
            unet = cls.load_model(unet_path, weight_quantization_config )
```

Comment (AlexKoff88) [inline optimum/intel/openvino/modeling_diffusion.py:295]:
```suggestion
            weight_quantization_params = {param: value for param, value in quantization_config.__dict__.items() if param != "dataset"}
            weight_quantization_config = OVWeightQuantizationConfig.from_dict(weight_quantization_params)
        else:
            weight_quantization_config = quantization_config
            unet = cls.load_model(unet_path, weight_quantization_config)
```

**Review by AlexKoff88 — APPROVED** [review]

**Review by AlexKoff88 — APPROVED** [review]

**Comment by AlexKoff88** [comment]
@echarlaix, PR is ready and we will update the SD example from QAT to the new approach in the follow-up PR.

**Review by echarlaix — APPROVED** [review]
Looks great thanks @l-bat
